bigquery/src/main/java/org/apache/iceberg/gcp/bigquery/metastore/BigQueryMetastoreCatalog.java

```diff
@@ -0,0 +1,328 @@
+  }
+
+  @Override
+  public boolean dropTable(TableIdentifier identifier, boolean purge) {
+    try {
+      TableOperations ops = newTableOps(identifier);
+      TableMetadata lastMetadata = ops.current();
+
+      client.deleteTable(toBqTableReference(identifier));
+
+      if (purge && lastMetadata != null) {
+        CatalogUtil.dropTableData(ops.io(), lastMetadata);
+      }
+
+    } catch (NoSuchTableException e) { // Not catching a NoSuchIcebergTableException on purpose
+      return false; // The documentation says just return false in this case
+    }
+
+    return true;
+  }
+
+  @Override
+  public void renameTable(TableIdentifier from, TableIdentifier to) {
+    if (!from.namespace().equals(to.namespace())) {
+      throw new ValidationException("New table name must be in the same namespace");
+    }
+
+    // TODO(b/354981675): Enable once supported by the API.
+    throw new ServiceFailureException(
+        "Table rename operation is unsupported. Try the SQL operation directly on BigQuery: \"ALTER TABLE %s RENAME TO %s;\"",
+        from.name(), to.name());
+  }
+
+  @Override
+  public void createNamespace(Namespace namespace, Map<String, String> metadata) {
+    Dataset builder = new Dataset();
+    DatasetReference datasetReference = toDatasetReference(namespace);
+    builder.setLocation(this.location);
+    builder.setDatasetReference(datasetReference);
+    builder.setExternalCatalogDatasetOptions(
+        BigQueryMetastoreUtils.createExternalCatalogDatasetOptions(
+            getDefaultStorageLocationUri(datasetReference.getDatasetId()), metadata));
+
+    client.createDataset(builder);
+  }
+
+  /**
+   * Since this catalog only supports one-level namespaces, it always returns an empty list unless
+   * passed an empty namespace to list all namespaces within the catalog.
+   */
+  @Override
+  public List<Namespace> listNamespaces(Namespace namespace) {
+    if (namespace.levels().length != 0) {
+      // BQMS does not support namespaces under database or tables, returns empty.
+      // It is called when dropping a namespace to make sure it's empty (listTables is called as
+      // well), returns empty to unblock deletion.
+      return ImmutableList.of();
+    }
+
+    return client.listDatasets(projectId).stream()
+        .map(BigQueryMetastoreCatalog::getNamespace)
+        .collect(ImmutableList.toImmutableList());
+  }
+
+  @Override
+  public boolean dropNamespace(Namespace namespace) {
+    client.deleteDataset(toDatasetReference(namespace));
+    /* We don't delete the data folder for safety, which aligns with Hive Metastore's default
+     * behavior.
+     * We can support database or catalog level config controlling file deletion in the future.
+     */
+    return true;
+  }
+
+  @Override
+  public boolean setProperties(Namespace namespace, Map<String, String> properties) {
+    client.setDatasetParameters(toDatasetReference(namespace), properties);
+    return true;
```

Comment:
These should return true if the properties are changed. Does this detect when the operation is a noop?